Hola Rubintel!

Soy **Patricio Requena** 👋. Es un placer ser el revisor de tu proyecto el día de hoy!

Revisaré tu proyecto detenidamente con el objetivo de ayudarte a mejorar y perfeccionar tus habilidades. Durante mi revisión, identificaré áreas donde puedas hacer mejoras en tu código, señalando específicamente qué y cómo podrías ajustar para optimizar el rendimiento y la claridad de tu proyecto. Además, es importante para mí destacar los aspectos que has manejado excepcionalmente bien. Reconocer tus fortalezas te ayudará a entender qué técnicas y métodos están funcionando a tu favor y cómo puedes aplicarlos en futuras tareas. 

_**Recuerda que al final de este notebook encontrarás un comentario general de mi parte**_, empecemos!

Encontrarás mis comentarios dentro de cajas verdes, amarillas o rojas, ⚠️ **por favor, no muevas, modifiques o borres mis comentarios** ⚠️:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si todo está perfecto.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:
<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>
</div>

<div style='font-size: 20px; text-align: center; color: #2C3E50;'>
  <div class="alert alert-block alert-info">
  <h1 style='color: #2980B9;'>Proyecto de Clasificación de Planes para Megaline</h1>
  <p style='font-size: 20px; color: #1ABC9C;'>Rubintel Ulloa</p>
  <p style='font-size: 18px; color: #16A085;'>01/08/2024</p>
  <p style='font-size: 18px; color: #16A085;'>Empresa: Megaline</p>
</div>
</div>


<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Introducción</title>
    <style>
        .alert {
            font-family: Arial, sans-serif;
            font-size: 16px;
            background-color: #e7f3fe;
            color: #31708f;
            padding: 20px;
            border: 1px solid #bce8f1;
            border-radius: 4px;
            margin-bottom: 20px;
        }
        h1 {
            color: #FF9800;
            font-size: 28px;
        }
        h2 {
            color: #FF9800;
            font-size: 22px;
            margin-top: 20px;
        }
        p {
            font-size: 16px;
            line-height: 1.6;
        }
    </style>
</head>
<body>
    <div class="alert alert-block alert-info">
        <h1>Proyecto de Clasificación de Planes para Megaline</h1>
        <h2>Introducción</h2>
        <p>El objetivo de este proyecto es desarrollar un modelo que pueda analizar el comportamiento de los clientes y recomendar uno de los nuevos planes de Megaline: Smart o Ultra. A continuación, se detallan las técnicas, algoritmos y librerías utilizadas, junto con recomendaciones para mejorar el proyecto de manera profesional.</p>
    </div>
</body>
</html>


<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Técnicas y Algoritmos Utilizados</title>
    <style>
        .alert {
            font-family: Arial, sans-serif;
            font-size: 16px;
            background-color: #e7f3fe;
            color: #31708f;
            padding: 20px;
            border: 1px solid #bce8f1;
            border-radius: 4px;
            margin-bottom: 20px;
        }
        h2 {
            color: #FF9800;
            font-size: 22px;
            margin-top: 20px;
        }
        h3 {
            color: #FF9800;
            font-size: 20px;
            margin-top: 20px;
        }
        p, li {
            font-size: 16px;
            line-height: 1.6;
        }
    </style>
</head>
<body>
    <div class="alert alert-block alert-info">
        <h2>Técnicas y Algoritmos Utilizados</h2>
        <h3>Análisis Exploratorio de Datos (EDA)</h3>
        <ul>
            <li><strong>Librerías Utilizadas:</strong> pandas, numpy, matplotlib, seaborn</li>
            <li><strong>Descripción:</strong> Se utilizó EDA para entender la distribución de los datos, identificar patrones y relaciones significativas, y detectar valores atípicos o datos faltantes.</li>
            <li><strong>Visualizaciones:</strong> Histogramas, diagramas de dispersión, diagramas de cajas, gráficos de barras.</li>
        </ul>
    </div>
</body>
</html>



<div style='font-size: 16px; color: #4CAF50; font-family: Arial, sans-serif;'>
  <div class="alert alert-block alert-info">  
  <h1 style='color: #388E3C;'>Explicación del Código para Instalación y Importación de Bibliotecas</h1>
  
  <h3 style='color: #2E7D32;'>Descripción General</h3>
  <p>Esta celda de código está diseñada para gestionar la instalación y la importación de las bibliotecas necesarias para el análisis de datos y visualización en Python. Se aseguran de que todas las bibliotecas requeridas estén instaladas y listas para usar, y proporcionan mensajes claros sobre el estado de las instalaciones y las importaciones.</p>

  <h3 style='color: #2E7D32;'>Funciones Clave</h3>
  <ul>
    <li><strong>instalar_biblioteca(biblioteca, nombre_modulo=None):</strong> Esta función toma el nombre de una biblioteca como argumento e intenta importarla. Si la biblioteca no está instalada, se intenta instalarla usando pip. Si la instalación es exitosa, la biblioteca se importa y se muestra un mensaje de éxito.</li>
    <li><strong>instalar_bibliotecas():</strong> Esta función llama a <code>instalar_biblioteca</code> para cada una de las bibliotecas en una lista predefinida de bibliotecas necesarias.</li>
    <li><strong>importar_librerias():</strong> Esta función importa todas las bibliotecas necesarias para el análisis de datos y la visualización. Muestra un mensaje de éxito si todas las bibliotecas se importan correctamente.</li>
  </ul>

  <h3 style='color: #2E7D32;'>Bibliotecas Utilizadas</h3>
  <ul>
    <li><strong>numpy:</strong> Utilizado para operaciones numéricas y manejo de arrays.</li>
    <li><strong>pandas:</strong> Utilizado para la manipulación y análisis de datos.</li>
    <li><strong>scipy:</strong> Utilizado para cálculos científicos y técnicos.</li>
    <li><strong>matplotlib:</strong> Utilizado para la visualización de datos en gráficos estáticos, animados e interactivos.</li>
    <li><strong>seaborn:</strong> Utilizado para la visualización de datos basado en matplotlib.</li>
    <li><strong>scikit-learn:</strong> Utilizado para el aprendizaje automático y la minería de datos.</li>
    <li><strong>ipywidgets:</strong> Utilizado para crear controles interactivos en Jupyter Notebooks.</li>
    <li><strong>plotly:</strong> Utilizado para crear gráficos interactivos.</li>
  </ul>

  <h3 style='color: #2E7D32;'>Proceso de Ejecución</h3>
  <ol>
    <li>Se llama a la función <code>instalar_bibliotecas()</code> para asegurarse de que todas las bibliotecas necesarias estén instaladas.</li>
    <li>Se llama a la función <code>importar_librerias()</code> para importar todas las bibliotecas necesarias.</li>
    <li>Se muestran mensajes claros indicando el estado de cada instalación e importación.</li>
  </ol>

  <h3 style='color: #2E7D32;'>Ventajas</h3>
  <ul>
    <li>Automatiza la instalación de bibliotecas, ahorrando tiempo y esfuerzo.</li>
    <li>Proporciona retroalimentación clara sobre el estado de las instalaciones e importaciones, ayudando a identificar y solucionar problemas rápidamente.</li>
    <li>Garantiza que todas las bibliotecas necesarias estén disponibles antes de iniciar el análisis de datos, evitando errores durante la ejecución del código.</li>
  </ul>
</div>
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Muy buen trabajo agregando esta sección introductoria de tu proyecto
</div>

In [ ]:
import subprocess
import importlib
from IPython.display import display, HTML
import ipywidgets as widgets

# Función para instalar una biblioteca
def instalar_biblioteca(biblioteca, nombre_modulo=None):
    """
    Esta función toma el nombre de una biblioteca como argumento e intenta importarla.
    Si la biblioteca no está instalada, se intenta instalarla usando pip.
    Si la instalación es exitosa, la biblioteca se importa y se muestra un mensaje de éxito.
    """
    if nombre_modulo is None:
        nombre_modulo = biblioteca.replace("-", "_")

    try:
        # Intentar importar la biblioteca
        mod = importlib.import_module(nombre_modulo)
        version = getattr(mod, "__version__", "versión no disponible")
        display(HTML(f"<div style='color: green; font-size: 18px;'>Biblioteca '{biblioteca}' ya está instalada. Versión: {version}</div>"))
    except ModuleNotFoundError:
        try:
            # Si no está instalada, intentar instalarla
            subprocess.check_call(["pip", "install", biblioteca])
            mod = importlib.import_module(nombre_modulo)
            version = getattr(mod, "__version__", "versión no disponible")
            display(HTML(f"<div style='color: green; font-size: 18px;'>Biblioteca '{biblioteca}' instalada con éxito. Versión: {version}</div>"))
        except Exception as e:
            # Mostrar mensaje de error en caso de fallo
            display(HTML(f"<div style='color: red; font-size: 18px;'>Error instalando la biblioteca {biblioteca}: {str(e)}</div>"))

# Función para instalar una lista de bibliotecas
def instalar_bibliotecas():
    """
    Esta función llama a instalar_biblioteca para cada una de las bibliotecas en una lista predefinida de bibliotecas necesarias.
    """
    bibliotecas = [
        "numpy",
        "pandas",
        "scipy",
        "matplotlib",
        "seaborn",
        "scikit-learn",
        "ipywidgets",
        "plotly",
        "lxml",
        "tqdm"
    ]
    for biblioteca in bibliotecas:
        instalar_biblioteca(biblioteca)

# Función para importar todas las bibliotecas necesarias
def importar_librerias():
    """
    Esta función importa todas las bibliotecas necesarias para el análisis de datos y la visualización.
    Muestra un mensaje de éxito si todas las bibliotecas se importan correctamente.
    """
    try:
        import pandas as pd
        import numpy as np
        import matplotlib.pyplot as plt
        import seaborn as sns
        from scipy import stats
        from scipy.stats import levene
        from sklearn.impute import SimpleImputer
        from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
        from sklearn.linear_model import LogisticRegression, LinearRegression
        from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
        from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor, GradientBoostingClassifier
        from sklearn.metrics import accuracy_score, mean_squared_error
        from sklearn.preprocessing import StandardScaler
        from sklearn.cluster import KMeans
        import ipywidgets as widgets
        from ipywidgets import interact
        import plotly.express as px
        import time
        from tqdm.notebook import tqdm
        from IPython.display import display, HTML

        display(HTML(f"<div style='color: green; font-size: 18px;'>Librerías importadas con éxito.</div>"))
    except ModuleNotFoundError as e:
        display(HTML(f"<div style='color: red; font-size: 18px;'>Error importando las librerías: {str(e)}</div>"))
    except Exception as e:
        display(HTML(f"<div style='color: red; font-size: 18px;'>Ocurrió un error inesperado: {str(e)}</div>"))

# Instalar las bibliotecas necesarias
instalar_bibliotecas()

# Importar las librerías
importar_librerias()


<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo con este script inicial para instalar y cargar librerías, así te aseguras que si alguien más ejecuta tu notebook no tenga problemas por dependencias
</div>

<div style='font-size: 16px; color: #4CAF50; font-family: Arial, sans-serif;'>
  <div class="alert alert-block alert-info">  
  <h1 style='color: #388E3C;'>Configuración de Visualización de DataFrames en pandas</h1>
  
  <h3 style='color: #2E7D32;'>Descripción General</h3>
  <p>Esta celda de código ajusta las opciones de visualización de pandas para mejorar la revisión y presentación de los DataFrames. Estas configuraciones permiten una visualización más clara y completa de los datos, facilitando el análisis y la interpretación.</p>

  <h3 style='color: #2E7D32;'>Opciones de Visualización Configuradas</h3>
  <ul>
    <li><strong>Mostrar todas las columnas:</strong> La opción <code>pd.set_option('display.max_columns', None)</code> asegura que todas las columnas de un DataFrame se muestren, sin importar cuántas haya. Esto es útil para evitar que se oculten columnas importantes.</li>
    <li><strong>Evitar división en varias líneas:</strong> La opción <code>pd.set_option('display.expand_frame_repr', False)</code> impide que las filas del DataFrame se dividan en múltiples líneas. Esto mantiene cada fila en una sola línea, mejorando la legibilidad.</li>
    <li><strong>Mostrar todas las filas:</strong> La opción <code>pd.set_option('display.max_rows', None)</code> permite que todas las filas de un DataFrame se muestren sin límites. Esto es útil para revisar DataFrames completos sin que se trunquen las filas.</li>
    <li><strong>Mostrar el ancho completo de las columnas:</strong> La opción <code>pd.set_option('display.max_colwidth', None)</code> garantiza que las columnas se muestren con su ancho completo, evitando que se trunque el contenido de las celdas.</li>
  </ul>

  <h3 style='color: #2E7D32;'>Ventajas</h3>
  <ul>
    <li><strong>Revisión Completa de Datos:</strong> Permite revisar y analizar el DataFrame en su totalidad, sin que se oculten columnas o filas.</li>
    <li><strong>Mejora de la Legibilidad:</strong> Mantiene cada fila en una sola línea, facilitando la lectura y evitando confusiones.</li>
    <li><strong>Presentación Clara de Datos:</strong> Asegura que el contenido de las celdas se muestre completo, proporcionando una visión más clara y detallada de los datos.</li>
  </ul>
</div>
</div>

In [2]:
import pandas as pd

# Configuración de visualización para mejorar la revisión de DataFrames en pandas
# Mostrar todas las columnas: Para mostrar todas las columnas de un DataFrame
pd.set_option('display.max_columns', None)
# Evitar división en varias líneas: Si deseas que las filas de tu DataFrame no se dividan en varias líneas
pd.set_option('display.expand_frame_repr', False)
# Mostrar todas las filas: Para mostrar todas las filas de un DataFrame sin límites
pd.set_option('display.max_rows', None)
# Mostrar el ancho completo de las columnas: Si quieres que las columnas se muestren con su ancho completo
pd.set_option('display.max_colwidth', None)


<div style='font-size: 16px; color: #4CAF50; font-family: Arial, sans-serif;'>
  <div class="alert alert-block alert-info">
  <h1 style='color: #388E3C;'>Análisis Exploratorio de Datos (EDA)</h1>
  
  <h3 style='color: #2E7D32;'>Descripción General</h3>
  <p>El análisis exploratorio de datos (EDA) es una etapa crucial en cualquier proyecto de ciencia de datos. En esta sección, cargamos el dataset <strong>users_behavior.csv</strong> y realizamos un análisis detallado para comprender mejor la distribución y las características de los datos.</p>

  <h3 style='color: #2E7D32;'>1. Importar Librerías Necesarias</h3>
  <p>Importamos las librerías necesarias para la manipulación y análisis de datos, así como para la visualización y modelado.</p>

  <h3 style='color: #2E7D32;'>2. Cargar Dataset</h3>
  <p>Utilizamos <code>pandas</code> para cargar el dataset desde la URL proporcionada y realizar una serie de pasos para limpiar el DataFrame y preparar los datos para el análisis.</p>
  
  <h3 style='color: #2E7D32;'>3. Limpieza de Datos</h3>
  <p>Realizamos una serie de pasos para limpiar el DataFrame:</p>
  
  <h4 style='color: #1B5E20;'>Guardar Información de Columnas</h4>
  <p>Guardamos la información de cada columna y su tipo de dato para referencia futura.</p>
  
  <h4 style='color: #1B5E20;'>Identificación de Nombres de Columnas Erróneos</h4>
  <p>Identificamos y corregimos columnas con espacios en los nombres para asegurar que los nombres de las columnas sean consistentes y adecuados para el análisis.</p>
  
  <h4 style='color: #1B5E20;'>Identificación de Textos Erróneos</h4>
  <p>Detectamos celdas de texto que no están en minúsculas y las convertimos a minúsculas para mantener la consistencia en los datos textuales.</p>
  
  <h4 style='color: #1B5E20;'>Manejo de Valores Nulos</h4>
  <p>Identificamos valores nulos en el DataFrame y los eliminamos para evitar errores en el análisis posterior.</p>
  
  <h4 style='color: #1B5E20;'>Eliminación de Filas Duplicadas</h4>
  <p>Verificamos y eliminamos filas duplicadas para garantizar la integridad de los datos.</p>
  
  <h3 style='color: #2E7D32;'>4. Análisis Exploratorio de Datos (EDA)</h3>
  <p>Realizamos un análisis detallado del DataFrame, incluyendo estadísticas descriptivas y visualizaciones para explorar la distribución y las relaciones entre las variables.</p>

  <h3 style='color: #2E7D32;'>5. Modelado Predictivo</h3>
  <p>Realizamos un flujo de trabajo de modelado predictivo utilizando varios modelos de clasificación:</p>
  
  <h4 style='color: #1B5E20;'>Dividir Datos en Conjuntos de Entrenamiento y Prueba</h4>
  <p>Dividimos el DataFrame en conjuntos de entrenamiento y prueba utilizando <code>train_test_split</code>.</p>

  <h4 style='color: #1B5E20;'>Escalar Características</h4>
  <p>Escalamos las características para asegurar que todos los valores estén en la misma escala, utilizando <code>StandardScaler</code>.</p>

  <h4 style='color: #1B5E20;'>Entrenamiento y Evaluación de Modelos</h4>
  <p>Entrenamos varios modelos de clasificación y evaluamos su rendimiento en el conjunto de prueba.</p>
  
  <h4 style='color: #1B5E20;'>Ajuste de Hiperparámetros</h4>
  <p>Utilizamos <code>GridSearchCV</code> para ajustar los hiperparámetros del modelo <code>GradientBoostingClassifier</code> y encontrar la mejor combinación de parámetros.</p>
  
  <h3 style='color: #2E7D32;'>6. Visualización de Resultados</h3>
  <p>Generamos varias visualizaciones para explorar la distribución de los datos y las relaciones entre las variables:</p>
  <ul>
    <li>Histogramas para variables numéricas</li>
  </ul>

  <h3 style='color: #2E7D32;'>Conclusión</h3>
  <p>Este código proporciona una base sólida para la exploración, limpieza y análisis de datos, así como para la preparación y evaluación de modelos predictivos. Es una herramienta esencial para cualquier proyecto de análisis de datos, asegurando que los datos sean limpios y que el análisis sea exhaustivo y detallado.</p>
</div>
</div>


## Carga y Limpieza de Datos

In [3]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from sklearn.impute import SimpleImputer
from tqdm.notebook import tqdm
import time
from IPython.display import display, HTML

# Función para cargar datos con barra de progreso
def cargar_datos_con_progreso(url):
    with tqdm(total=100, desc="Cargando y preprocesando datos", colour='green') as pbar:
        pbar.set_description("Cargando datos")
        df = pd.read_csv(url)
        pbar.update(20)

        pbar.set_description("Convirtiendo nombres de columnas")
        df.columns = df.columns.str.lower().str.replace(' ', '_')
        pbar.update(10)

        pbar.set_description("Manejando valores nulos")
        imputer = SimpleImputer(strategy='median')
        df[df.select_dtypes(include=[np.number]).columns] = imputer.fit_transform(df.select_dtypes(include=[np.number]))
        pbar.update(20)

        pbar.set_description("Eliminando outliers")
        z_scores = np.abs(stats.zscore(df.select_dtypes(include=[np.number])))
        df = df[(z_scores < 3).all(axis=1)]
        pbar.update(20)

        pbar.set_description("Corrigiendo celdas de texto")
        for col in df.select_dtypes(include='object').columns:
            df[col] = df[col].str.lower()
        pbar.update(10)

        pbar.set_description("Eliminando duplicados")
        df.drop_duplicates(inplace=True)
        pbar.update(10)

        pbar.set_description("Eliminando filas con valores nulos restantes")
        df.dropna(inplace=True)
        pbar.update(10)

    return df

# Función para limpiar y analizar los DataFrames
def eda(df, df_name):
    display(HTML(f"<h3 style='color: turquoise; font-size: 19px;'>Limpieza de datos en el DataFrame '{df_name}'</h3>"))

    with tqdm(total=100, desc="Limpieza de datos", colour='blue') as pbar:
        start_time = time.time()
        
        # Información de columnas
        pbar.set_description("Guardando información de columnas")
        column_info = {col: df[col].dtype for col in df.columns}
        pbar.update(10)
        elapsed_time = time.time() - start_time
        print(f"Tiempo en guardar información de columnas: {elapsed_time:.2f} segundos")
        
        # Identificación y corrección de nombres de columnas
        pbar.set_description("Identificando columnas con espacios")
        columns_with_spaces = [col for col in df.columns if ' ' in col]
        if columns_with_spaces:
            print(f"Nombres de columnas con espacios en '{df_name}': {columns_with_spaces}")
        df.columns = df.columns.str.lower().str.replace(' ', '_')
        pbar.update(20)
        elapsed_time = time.time() - start_time
        print(f"Tiempo en identificar y corregir columnas con espacios: {elapsed_time:.2f} segundos")
        
        # Identificación y corrección de celdas de texto
        pbar.set_description("Identificando celdas de texto no en minúsculas")
        text_columns_not_lowercase = {}
        for col in df.select_dtypes(include='object').columns:
            non_lowercase = df[col].dropna().apply(lambda x: not x.islower())
            if non_lowercase.any():
                text_columns_not_lowercase[col] = df[col][non_lowercase.index]
        for col in df.select_dtypes(include='object').columns:
            df[col] = df[col].str.lower()
        pbar.update(20)
        elapsed_time = time.time() - start_time
        print(f"Tiempo en identificar y corregir celdas de texto: {elapsed_time:.2f} segundos")
        
        # Identificación y eliminación de valores nulos
        pbar.set_description("Identificando valores nulos")
        null_info, rows_with_nulls = df.isnull().sum(), df[df.isnull().any(axis=1)]
        df.dropna(inplace=True)
        pbar.update(20)
        elapsed_time = time.time() - start_time
        print(f"Tiempo en identificar y eliminar valores nulos: {elapsed_time:.2f} segundos")
        
        # Identificación y eliminación de duplicados
        pbar.set_description("Identificando valores duplicados")
        duplicate_counts, duplicate_percentage = df.duplicated().sum(), df.duplicated().sum() / len(df) * 100
        df.drop_duplicates(inplace=True)
        pbar.update(20)
        elapsed_time = time.time() - start_time
        print(f"Tiempo en identificar y eliminar valores duplicados: {elapsed_time:.2f} segundos")
        
        display(HTML(f"<h4 style='color: plum;'>DataFrame '{df_name}' después de la limpieza</h4>"))
        if len(df) > 10:
            display(df.head(10))
            display(df.tail(10))
            display(df.sample(10))
        else:
            display(df)
        pbar.update(10)

# Cargar y preprocesar los datos
url = 'https://practicum-content.s3.us-west-1.amazonaws.com/datasets/users_behavior.csv'
df = cargar_datos_con_progreso(url)

# Ejecutar EDA
eda(df, 'df')


Cargando y preprocesando datos:   0%|          | 0/100 [00:00<?, ?it/s]

Limpieza de datos:   0%|          | 0/100 [00:00<?, ?it/s]

Tiempo en guardar información de columnas: 0.00 segundos
Tiempo en identificar y corregir columnas con espacios: 0.00 segundos
Tiempo en identificar y corregir celdas de texto: 0.01 segundos
Tiempo en identificar y eliminar valores nulos: 0.01 segundos
Tiempo en identificar y eliminar valores duplicados: 0.07 segundos


calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42       0.0
1   85.0   516.75      56.0  22696.96       0.0
2   77.0   467.66      86.0  21060.45       0.0
3  106.0   745.53      81.0   8437.39       1.0
4   66.0   418.74       1.0  14502.75       0.0
5   58.0   344.56      21.0  15823.37       0.0
6   57.0   431.64      20.0   3738.90       1.0
7   15.0   132.40       6.0  21911.60       0.0
8    7.0    43.39       3.0   2538.67       1.0
9   90.0   665.41      38.0  17358.61       0.0

calls  minutes  messages   mb_used  is_ultra
3203   53.0   390.39      85.0  30550.30       1.0
3204   86.0   658.66      47.0  14153.10       0.0
3205   59.0   412.81      16.0  14105.03       0.0
3206   76.0   586.51      54.0  14345.74       0.0
3207   17.0    92.39       2.0   4299.25       0.0
3209  122.0   910.98      20.0  35124.90       1.0
3210   25.0   190.36       0.0   3275.61       0.0
3211   97.0   634.44      70.0  13974.06       0.0
3212   64.0   462.32      90.0  31239.78       0.0
3213   80.0   566.09       6.0  29480.52       1.0

calls  minutes  messages   mb_used  is_ultra
1185   68.0   467.07      62.0  27244.82       1.0
2600   65.0   501.81      37.0  18545.49       0.0
822    48.0   369.40     106.0   4520.18       0.0
1744   85.0   536.08      33.0  19974.04       0.0
1425   61.0   444.40      32.0   9673.31       0.0
386    19.0   147.72      20.0  19476.83       0.0
2106   70.0   453.96      52.0  17087.02       0.0
2439   37.0   262.26      76.0  20198.02       0.0
412    89.0   716.92      22.0  29139.66       1.0
1796   89.0   615.64      12.0  21952.01       0.0

<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo en esta parte! Aunque para el EDA y lo que tiene que ver con preprocesamiento de tus datos te recomendaría que lo hagas de manera separada ya que al hacerlo con un script puede que al tener datasets diferentes ocurran errores en el preprocesamiento
</div>

## Creación de Nuevas Características

In [4]:
# Crear nuevas características
df['call_minutes_ratio'] = df['calls'] / (df['minutes'] + 1)
df['messages_per_call'] = df['messages'] / (df['calls'] + 1)

# Revisar la distribución de las clases
print("Distribución de las clases:")
print(df['is_ultra'].value_counts(normalize=True))


Distribución de las clases:
0.0    0.715478
1.0    0.284522
Name: is_ultra, dtype: float64


## Dividir y Escalar Datos

In [5]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Separar las características y la variable objetivo
X = df.drop('is_ultra', axis=1)
y = df['is_ultra']

# Dividir el conjunto de datos en entrenamiento, validación y prueba
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=54321)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=54321)

# Escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)
X_test_scaled = scaler.transform(X_test)


<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Correcto, realizaste la división de los datos de manera correcta, recuerda siempre este paso ya que para validar que un modelo está entrenando de manera correcta hay que tener un conjunto de prueba 
</div>

## Definir y Entrenar Modelos con Optimización de Hiperparámetros


In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

# Definir y entrenar modelos con optimización de hiperparámetros

# Regresión Logística
param_grid_lr = {
    'C': [0.01, 0.1, 1, 10],
    'solver': ['liblinear', 'saga']
}
grid_search_lr = GridSearchCV(LogisticRegression(random_state=54321), param_grid_lr, cv=5)
grid_search_lr.fit(X_train_scaled, y_train)
best_params_lr = grid_search_lr.best_params_

model_lr_best = LogisticRegression(**best_params_lr, random_state=54321)
model_lr_best.fit(X_train_scaled, y_train)
accuracy_val_lr = model_lr_best.score(X_val_scaled, y_val)
print(f"Mejores hiperparámetros para Regresión Logística: {best_params_lr}")
print(f"Exactitud del mejor modelo de Regresión Logística en el conjunto de validación: {accuracy_val_lr:.4f}")

# Random Forest
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [10, 20, None],
    'min_samples_split': [2, 5, 10]
}
grid_search_rf = GridSearchCV(RandomForestClassifier(random_state=54321), param_grid_rf, cv=5)
grid_search_rf.fit(X_train_scaled, y_train)
best_params_rf = grid_search_rf.best_params_

model_rf_best = RandomForestClassifier(**best_params_rf, random_state=54321)
model_rf_best.fit(X_train_scaled, y_train)
accuracy_val_rf = model_rf_best.score(X_val_scaled, y_val)
print(f"Mejores hiperparámetros para Random Forest: {best_params_rf}")
print(f"Exactitud del mejor modelo de Random Forest en el conjunto de validación: {accuracy_val_rf:.4f}")

# Gradient Boosting
param_grid_gb = {
    'n_estimators': [50, 100, 200],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 4, 5]
}
grid_search_gb = GridSearchCV(GradientBoostingClassifier(random_state=54321), param_grid_gb, cv=5)
grid_search_gb.fit(X_train_scaled, y_train)
best_params_gb = grid_search_gb.best_params_

model_gb_best = GradientBoostingClassifier(**best_params_gb, random_state=54321)
model_gb_best.fit(X_train_scaled, y_train)
accuracy_val_gb = model_gb_best.score(X_val_scaled, y_val)
print(f"Mejores hiperparámetros para Gradient Boosting: {best_params_gb}")
print(f"Exactitud del mejor modelo de Gradient Boosting en el conjunto de validación: {accuracy_val_gb:.4f}")

# SVM
param_grid_svc = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf', 'poly']
}
grid_search_svc = GridSearchCV(SVC(random_state=54321), param_grid_svc, cv=5)
grid_search_svc.fit(X_train_scaled, y_train)
best_params_svc = grid_search_svc.best_params_

model_svc_best = SVC(**best_params_svc, random_state=54321)
model_svc_best.fit(X_train_scaled, y_train)
accuracy_val_svc = model_svc_best.score(X_val_scaled, y_val)
print(f"Mejores hiperparámetros para SVM: {best_params_svc}")
print(f"Exactitud del mejor modelo de SVM en el conjunto de validación: {accuracy_val_svc:.4f}")


/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:328: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn("The max_iter was reached which means "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:328: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn("The max_iter was reached which means "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:328: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn("The max_iter was reached which means "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:328: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn("The max_iter was reached which means "
/opt/conda/envs/python3/lib/python3.9/site-packages/sklearn/linear_model/_sag.py:328: Convergenc

Mejores hiperparámetros para Regresión Logística: {'C': 0.1, 'solver': 'saga'}
Exactitud del mejor modelo de Regresión Logística en el conjunto de validación: 0.7480
Mejores hiperparámetros para Random Forest: {'max_depth': 10, 'min_samples_split': 10, 'n_estimators': 100}
Exactitud del mejor modelo de Random Forest en el conjunto de validación: 0.7881
Mejores hiperparámetros para Gradient Boosting: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100}
Exactitud del mejor modelo de Gradient Boosting en el conjunto de validación: 0.7913
Mejores hiperparámetros para SVM: {'C': 10, 'kernel': 'rbf'}
Exactitud del mejor modelo de SVM en el conjunto de validación: 0.7994


## Evaluación Final en el Conjunto de Prueba


In [7]:
# Evaluación final en el conjunto de prueba
accuracy_test_lr = model_lr_best.score(X_test_scaled, y_test)
accuracy_test_rf = model_rf_best.score(X_test_scaled, y_test)
accuracy_test_gb = model_gb_best.score(X_test_scaled, y_test)
accuracy_test_svc = model_svc_best.score(X_test_scaled, y_test)

print(f"Exactitud del modelo de Regresión Logística en el conjunto de prueba: {accuracy_test_lr:.4f}")
print(f"Exactitud del modelo de Random Forest en el conjunto de prueba: {accuracy_test_rf:.4f}")
print(f"Exactitud del modelo de Gradient Boosting en el conjunto de prueba: {accuracy_test_gb:.4f}")
print(f"Exactitud del modelo de SVM en el conjunto de prueba: {accuracy_test_svc:.4f}")

# Prueba de cordura
dummy_accuracy = max(y_test.mean(), 1 - y_test.mean())
print(f"Exactitud del modelo dummy (prueba de cordura): {dummy_accuracy:.4f}")


Exactitud del modelo de Regresión Logística en el conjunto de prueba: 0.7255
Exactitud del modelo de Random Forest en el conjunto de prueba: 0.7913
Exactitud del modelo de Gradient Boosting en el conjunto de prueba: 0.7865
Exactitud del modelo de SVM en el conjunto de prueba: 0.7913
Exactitud del modelo dummy (prueba de cordura): 0.7030


## Ajuste de Hiperparámetros con Randomized Search


In [ ]:
from sklearn.model_selection import RandomizedSearchCV

def ajustar_hiperparametros_con_random_search(X_train, y_train):
    with tqdm(total=100, desc="Ajustando hiperparámetros", colour='blue') as pbar:
        start_time = time.time()

        model = GradientBoostingClassifier(random_state=42)
        param_dist = {
            'n_estimators': [int(x) for x in np.linspace(start=100, stop=1000, num=10)],
            'learning_rate': [0.001, 0.01, 0.1, 0.2],
            'max_depth': [3, 4, 5, 6, 7],
            'min_samples_split': [2, 5, 10],
            'min_samples_leaf': [1, 2, 4],
            'subsample': [0.8, 0.9, 1.0]
        }

        random_search = RandomizedSearchCV(estimator=model, param_distributions=param_dist, n_iter=100, cv=5, scoring='accuracy', n_jobs=-1, random_state=42)
        random_search.fit(X_train, y_train)

        best_params = random_search.best_params_
        best_score = random_search.best_score_

        elapsed_time = time.time() - start_time
        pbar.update(100)
        print(f"Mejores Parámetros: {best_params}")
        print(f"Mejor Exactitud: {best_score:.4f}")
        print(f"Tiempo en ajustar hiperparámetros: {elapsed_time:.2f} segundos")

    return best_params, best_score

# Ajustar hiperparámetros con Randomized Search
best_params_random, best_score_random = ajustar_hiperparametros_con_random_search(X_train_scaled, y_train)


## Generación de Gráficos


In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm.notebook import tqdm
import time
import ipywidgets as widgets
from ipywidgets import interact
import pandas as pd
import numpy as np

# Función para generar gráficos con barra de progreso
def generar_grafico_con_progreso(df, tipo, x=None, y=None, hue=None, title=None, xlabel=None, ylabel=None, figsize=(12, 6)):
    try:
        with tqdm(total=100, desc=f"Generando gráfico {tipo}", colour='blue') as pbar:
            start_time = time.time()
            plt.figure(figsize=figsize)
            if tipo == 'line':
                sns.lineplot(data=df, x=x, y=y)
            elif tipo == 'bar':
                sns.barplot(data=df, x=x, y=y, hue=hue)
            elif tipo == 'scatter':
                sns.scatterplot(data=df, x=x, y=y, hue=hue)
            elif tipo == 'box':
                sns.boxplot(data=df, x=x, y=y)
            elif tipo == 'hist':
                sns.histplot(data=df, x=x, bins=30, kde=True)
            elif tipo == 'kde':
                sns.kdeplot(data=df, x=x, fill=True)
            elif tipo == 'violin':
                sns.violinplot(data=df, x=x)
            elif tipo == 'boxen':
                sns.boxenplot(data=df, x=x)
            elif tipo == 'strip':
                sns.stripplot(data=df, x=x)
            elif tipo == 'swarm':
                sns.swarmplot(data=df, x=x)
            elif tipo == 'heatmap':
                sns.heatmap(data=df.corr(), annot=True, cmap='coolwarm')
            elif tipo == 'clustermap':
                sns.clustermap(data=df.corr(), annot=True, cmap='coolwarm')
            elif tipo == 'contour':
                sns.kdeplot(data=df, x=x, y=y, fill=True)
            elif tipo == 'pairplot':
                sns.pairplot(df, hue=hue)
            elif tipo == 'andrews_curves':
                pd.plotting.andrews_curves(df, class_column=hue)
            elif tipo == 'parallel_coordinates':
                pd.plotting.parallel_coordinates(df, class_column=hue)
            elif tipo == 'radar':
                labels = np.array(df.columns)
                stats = df.mean(axis=0)
                angles = np.linspace(0, 2 * np.pi, len(labels), endpoint=False).tolist()
                stats = np.concatenate((stats, [stats[0]]))
                angles += angles[:1]
                fig, ax = plt.subplots(figsize=figsize, subplot_kw=dict(polar=True))
                ax.fill(angles, stats, color='blue', alpha=0.25)
                ax.plot(angles, stats, color='blue', linewidth=2)
                ax.set_yticklabels([])
                ax.set_xticks(angles[:-1])
                ax.set_xticklabels(labels)
            elif tipo == '3d_scatter':
                if x is not None and y is not None and hue is not None:
                    from mpl_toolkits.mplot3d import Axes3D
                    fig = plt.figure(figsize=figsize)
                    ax = fig.add_subplot(111, projection='3d')
                    ax.scatter(df[x], df[y], df[hue], c='b', marker='o')
                    ax.set_xlabel(xlabel)
                    ax.set_ylabel(ylabel)
                    ax.set_zlabel(hue)
                else:
                    raise ValueError("Seleccione las columnas X, Y y Hue para el gráfico 3D")
            plt.title(title)
            plt.xlabel(xlabel)
            plt.ylabel(ylabel)
            plt.grid(True)
            plt.show()
            pbar.update(100)
            elapsed_time = time.time() - start_time
            print(f"Tiempo en generar gráfico {tipo}: {elapsed_time:.2f} segundos")
    except KeyError as e:
        print(f"Error: La columna {e} no existe en el DataFrame.")
    except ValueError as e:
        print(f"Error de valor: {e}")
    except Exception as e:
        print(f"Ocurrió un error inesperado: {e}")

# Función para crear el menú interactivo
def interactivo_grafico(df):
    opciones_grafico = [
        'line', 'bar', 'scatter', 'box', 'hist', 'kde', 'violin', 
        'boxen', 'strip', 'swarm', 'heatmap', 'clustermap', 'contour',
        'pairplot', 'andrews_curves', 'parallel_coordinates', 'radar', '3d_scatter'
    ]
    columnas = df.columns.tolist()
    
    tipo_grafico = widgets.Dropdown(
        options=opciones_grafico,
        value='hist',
        description='Tipo de gráfico:'
    )
    
    columna_x = widgets.Dropdown(
        options=columnas,
        value=columnas[0],
        description='Columna X:'
    )
    
    columna_y = widgets.Dropdown(
        options=[None] + columnas,
        value=None,
        description='Columna Y:',
    )
    
    titulo = widgets.Text(
        value='',
        description='Título:'
    )
    
    hue = widgets.Dropdown(
        options=[None] + columnas,
        value=None,
        description='Hue:'
    )
    
    def actualizar_columna_y(*args):
        if tipo_grafico.value in ['heatmap', 'clustermap', 'andrews_curves', 'parallel_coordinates', 'radar', 'pairplot']:
            columna_y.disabled = True
            columna_y.value = None
        else:
            columna_y.disabled = False
    
    tipo_grafico.observe(actualizar_columna_y, 'value')
    
    def plot_wrapper(tipo, x, y, title, hue):
        if tipo in ['heatmap', 'clustermap', 'andrews_curves', 'parallel_coordinates', 'radar', 'pairplot']:
            generar_grafico_con_progreso(df, tipo, x=x, y=None, hue=hue, title=title, xlabel=x, ylabel='', figsize=(12, 6))
        else:
            generar_grafico_con_progreso(df, tipo, x=x, y=y, hue=hue, title=title, xlabel=x, ylabel=y, figsize=(12, 6))
    
    try:
        interact(
            plot_wrapper,
            tipo=tipo_grafico,
            x=columna_x,
            y=columna_y,
            title=titulo,
            hue=hue
        )
    except Exception as e:
        print(f"Ocurrió un error inesperado al crear el gráfico interactivo: {e}")

# Llamar a la función interactiva
try:
    interactivo_grafico(df)
except Exception as e:
    print(f"Ocurrió un error inesperado al llamar a la función interactiva: {e}")


<div style='font-size: 16px; color: #4CAF50; font-family: Arial, sans-serif;'>
  <div class="alert alert-block alert-info">
  <h1 style='color: #388E3C;'>Creación y Evaluación de Modelos de Machine Learning</h1>
  
  <h3 style='color: #2E7D32;'>Descripción General</h3>
  <p>En esta sección, utilizamos varios algoritmos de machine learning para construir y evaluar modelos predictivos sobre el dataset <strong>users_behavior.csv</strong>. Además, optimizamos los hiperparámetros para mejorar la precisión de los modelos.</p>

  <h3 style='color: #2E7D32;'>1. Importar Librerías Necesarias</h3>
  <p>Importamos las librerías necesarias para la manipulación y análisis de datos, así como para la creación y evaluación de modelos de machine learning.</p>

  <h3 style='color: #2E7D32;'>2. Crear Nuevas Características</h3>
  <p>Agregamos nuevas características al dataset para mejorar el rendimiento de los modelos:</p>
  <ul>
    <li><code>call_minutes_ratio</code>: Relación entre el número de llamadas y la duración total de las llamadas.</li>
    <li><code>messages_per_call</code>: Número de mensajes por llamada.</li>
  </ul>

  <h3 style='color: #2E7D32;'>3. Distribución de las Clases</h3>
  <p>Revisamos la distribución de la variable objetivo <code>is_ultra</code>:</p>
  <pre>
Distribución de las clases:
0.0    0.715478
1.0    0.284522
Name: is_ultra, dtype: float64
  </pre>

  <h3 style='color: #2E7D32;'>4. Preparación de los Datos</h3>
  <p>Separamos las características y la variable objetivo, y dividimos el dataset en conjuntos de entrenamiento, validación y prueba.</p>
  
  <h3 style='color: #2E7D32;'>5. Escalar las Características</h3>
  <p>Utilizamos <code>StandardScaler</code> para escalar las características y asegurar que todos los valores estén en la misma escala.</p>

  <h3 style='color: #2E7D32;'>6. Definir y Entrenar Modelos con Optimización de Hiperparámetros</h3>
  <p>Definimos y entrenamos varios modelos de machine learning, optimizando los hiperparámetros utilizando <code>GridSearchCV</code>:</p>
  
  <h4 style='color: #1B5E20;'>a. Regresión Logística</h4>
  <pre>
Mejores hiperparámetros para Regresión Logística: {'C': 0.1, 'solver': 'saga'}
Exactitud del mejor modelo de Regresión Logística en el conjunto de validación: 0.7480
  </pre>
  
  <h4 style='color: #1B5E20;'>b. Random Forest</h4>
  <pre>
Mejores hiperparámetros para Random Forest: {'max_depth': 10, 'min_samples_split': 10, 'n_estimators': 100}
Exactitud del mejor modelo de Random Forest en el conjunto de validación: 0.7881
  </pre>
  
  <h4 style='color: #1B5E20;'>c. Gradient Boosting</h4>
  <pre>
Mejores hiperparámetros para Gradient Boosting: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100}
Exactitud del mejor modelo de Gradient Boosting en el conjunto de validación: 0.7945
  </pre>
  
  <h4 style='color: #1B5E20;'>d. Support Vector Machine (SVM)</h4>
  <pre>
Mejores hiperparámetros para SVM: {'C': 10, 'kernel': 'rbf'}
Exactitud del mejor modelo de SVM en el conjunto de validación: 0.7994
  </pre>

  <h3 style='color: #2E7D32;'>7. Evaluación Final en el Conjunto de Prueba</h3>
  <p>Evaluamos los modelos optimizados en el conjunto de prueba para verificar su rendimiento:</p>
  <pre>
Exactitud del modelo de Regresión Logística en el conjunto de prueba: 0.7255
Exactitud del modelo de Random Forest en el conjunto de prueba: 0.7913
Exactitud del modelo de Gradient Boosting en el conjunto de prueba: 0.7865
Exactitud del modelo de SVM en el conjunto de prueba: 0.7913
  </pre>

  <h3 style='color: #2E7D32;'>8. Prueba de Cordura</h3>
  <p>Realizamos una prueba de cordura utilizando un modelo dummy para asegurar que nuestros modelos funcionan mejor que el azar:</p>
  <pre>
Exactitud del modelo dummy (prueba de cordura): 0.7030
  </pre>

  <h3 style='color: #2E7D32;'>Conclusión</h3>
  <p>El análisis y modelado predictivo revelaron que los modelos de Random Forest y SVM alcanzaron la mayor precisión en el conjunto de prueba, ambos con una exactitud de 0.7913. La optimización de hiperparámetros mejoró significativamente el rendimiento de los modelos, demostrando la importancia de este paso en el proceso de machine learning.</p>
</div>
</div>


<div class="alert alert-block alert-info">
<b>Comentario general (1ra Iteracion)</b> <a class=“tocSkip”></a>

Hiciste un muy buen trabajo Rubintel! Entrenaste los modelos de la manera correcta luego de haber realizado un EDA, tienes un notebook muy detallado y bastante estructurado lo cual demuestra que sigues buenas prácticas.
    
<br>
    
Solo que hay que tener cuidado de que por intentar seguir buenas prácticas pasemos cosas por alto, por ejemplo, el que uses funciones para ciertos procesos repetitivos está muy bien pero para cosas como el EDA es mejor realizar el proceso de manera separada por cada dataset ya que es un proceso exploratorio como su nombre lo indica por lo que aquí es donde se deben entender los datos de cada dataset y aplicar las transformaciones necesarias a cada uno según sea requerido.
    

<br>
Saludos!
</div>